anndata/utils.py

```diff
@@ -1,17 +1,40 @@
 import warnings
-from functools import wraps
+from functools import wraps, singledispatch
 from typing import Mapping, Any, Sequence, Union, Sized, Optional
 
 import pandas as pd
 import numpy as np
 
 from .logging import get_logger
-if False:
-    from .base import BoundRecArr  # noqa
 
 logger = get_logger(__name__)
 
 
+@singledispatch
+def convert_to_dict(obj) -> dict:
+    return dict(obj)
+
+@convert_to_dict.register(dict)
+def convert_to_dict_dict(obj: dict):
+    return obj
+
+@convert_to_dict.register(np.ndarray)
+def convert_to_dict_ndarray(obj: np.ndarray):
+    if obj.dtype.fields is None:
+        raise TypeError(
+            "Can only convert np.ndarray with compound dtypes to dict, passed "
+            "array had '{}'.".format(obj.dtype)
+        )
+    d = dict()
+    for k in obj.dtype.fields.keys():
+        d[k] = obj[k]
+    return d
+
+@convert_to_dict.register(type(None))  # This is required for python < 3.7
```

Comment:
You mean the part `(type(None))` is required for those versions, right?

I initially interpreted it as “registering the function for NoneType at all is required for python 3.6” and was like “wait, does `dict(None)`” work in 3.7? I think the comment is more confusing than enlightening :wink:

Reply:
Yeah, but I actually just learned it was required for a different reason than I thought, and `.register(type)` is required for all pre 3.7 versions. Now it's not even necessary I think.